service/src/test/java/bio/terra/workspace/app/configuration/external/controller/ControlledGcpResourceApiControllerConnectedTest.java

```diff
@@ -0,0 +1,82 @@
+package bio.terra.workspace.app.configuration.external.controller;
+
+import static bio.terra.workspace.common.utils.MockMvcUtils.createBigQueryDataset;
+import static bio.terra.workspace.common.utils.MockMvcUtils.createGcsBucket;
+import static bio.terra.workspace.common.utils.MockMvcUtils.deleteWorkspace;
+import static bio.terra.workspace.common.utils.MockMvcUtils.getBigQueryDataset;
+import static bio.terra.workspace.common.utils.MockMvcUtils.getGcsBucket;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import bio.terra.workspace.common.BaseConnectedTest;
+import bio.terra.workspace.connected.UserAccessUtils;
+import bio.terra.workspace.connected.WorkspaceConnectedTestUtils;
+import bio.terra.workspace.generated.model.ApiCreatedControlledGcpBigQueryDataset;
+import bio.terra.workspace.generated.model.ApiCreatedControlledGcpGcsBucket;
+import bio.terra.workspace.generated.model.ApiGcpBigQueryDatasetResource;
+import bio.terra.workspace.generated.model.ApiGcpGcsBucketResource;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.UUID;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.test.web.servlet.MockMvc;
+
+@AutoConfigureMockMvc
+public class ControlledGcpResourceApiControllerConnectedTest extends BaseConnectedTest {
+
+  @Autowired MockMvc mockMvc;
+  @Autowired ObjectMapper objectMapper;
+  @Autowired UserAccessUtils userAccessUtils;
+  @Autowired WorkspaceConnectedTestUtils connectedTestUtils;
+
+  private UUID workspaceId;
+
+  @BeforeEach
+  public void setUp() throws Exception {
+    workspaceId =
+        connectedTestUtils
+            .createWorkspaceWithGcpContext(userAccessUtils.defaultUserAuthRequest())
+            .getWorkspaceId();
+  }
+
+  @AfterEach
+  public void cleanup() throws Exception {
+    deleteWorkspace(workspaceId, mockMvc, userAccessUtils.defaultUserAuthRequest());
+  }
+
+  @Test
+  public void createControlledBigQueryDataset() throws Exception {
+    ApiCreatedControlledGcpBigQueryDataset resource =
+        createBigQueryDataset(
+            mockMvc, objectMapper, workspaceId, userAccessUtils.defaultUserAuthRequest());
+
+    ApiGcpBigQueryDatasetResource retrievedResource =
+        getBigQueryDataset(
+            mockMvc,
+            objectMapper,
+            workspaceId,
+            resource.getResourceId(),
+            userAccessUtils.defaultUserAuthRequest());
+
+    assertEquals(resource.getBigQueryDataset(), retrievedResource);
```

Comment:
In general, don't use response as expectation. Say there's a bug and createBigQueryDataset returns the wrong thing, eg properties is missing. This test won't catch that.

You could write assertBigQueryDataset() and call that.

Same for other test.

Reply:
why?

i don't follow. If there is a bug in create, wouldn't the get and create return the different things and get caught here.

My concern with asserting individual fields if someone adds a new field, they wouldn't know to come here to add a new assert line and it would be easily missed. And that would definitely let bug slips through.

Reply:
Say there's a bug in create and get where properties aren't returned. This `assertEquals` won't catch that, because both expected and actual are missing properties.

If someone adds a new field, it is their responsibility to have at least one test check that field. (They don't necessarily need to update all tests.)